src/lang/cpp/internal/TypeResolver.java

```diff
@@ -569,111 +621,135 @@ private IConstructor resolveICPPEnumerationSpecialization(ICPPEnumerationSpecial
-		if (type.isVolatile())
+		}
+		if (type.isVolatile()) {
 			modifiers.append(builder.TypeModifier_volatile());
-		if (type.isRestrict())
+		}
+		if (type.isRestrict()) {
 			modifiers.append(builder.TypeModifier_restrict());
-		return builder.TypeSymbol_pointerType(modifiers.done(), resolveType(type.getType()));
+		}
+		return builder.TypeSymbol_pointerType(modifiers.done(), resolveType(type.getType(), origin));
 	}
 
-	private IConstructor resolveIProblemBinding(IProblemBinding type) {
+	private IConstructor resolveIProblemBinding(IProblemBinding type, ISourceLocation origin) {
 		// err("Encountered IProblemBinding " + type.getClass().getSimpleName()
 		// + ": ");
 		// err("\t" + type.getID() + ": " + type.getMessage());
-		return builder.TypeSymbol_problemBinding();
+		return builder.TypeSymbol_problemBinding().asWithKeywordParameters().setParameter("src", origin);
 	}
 
 	private IConstructor resolveIProblemType(IProblemType type) {
 		return builder.TypeSymbol_problemType(type.getMessage());
 	}
 
-	private IConstructor resolveIQualifierType(IQualifierType type) {
-		IConstructor baseType = resolveType(type.getType());
+	private IConstructor resolveIQualifierType(IQualifierType type, ISourceLocation origin) {
+		IConstructor baseType = resolveType(type.getType(), origin);
 		IListWriter modifiers = vf.listWriter();
-		if (type.isConst())
+		if (type.isConst()) {
 			modifiers.append(builder.TypeModifier_const());
-		if (type.isVolatile())
+		}
+		if (type.isVolatile()) {
 			modifiers.append(builder.TypeModifier_volatile());
+		}
 		return builder.TypeSymbol_qualifierType(modifiers.done(), baseType);
 	}
 
-	private IConstructor resolveITypeContainer(ITypeContainer type) {
-		return builder.TypeSymbol_typeContainer(resolveType(type.getType()));
+	private IConstructor resolveITypeContainer(ITypeContainer type, ISourceLocation origin) {
+		return builder.TypeSymbol_typeContainer(resolveType(type.getType(), origin));
 	}
 
-	private IConstructor resolveITypedef(ITypedef type) {
-		if (type instanceof ICPPAliasTemplateInstance)
-			throw new RuntimeException("NYI: ICPPAliasTemplateInstance");
-		return builder.TypeSymbol_typedef(resolveType(type.getType()));
+	private IConstructor resolveITypedef(ITypedef type, ISourceLocation origin) {
+		if (type instanceof ICPPAliasTemplateInstance) {
+			ICPPAliasTemplateInstance ati = (ICPPAliasTemplateInstance) type;
+			IListWriter templateParameters = vf.listWriter();
+			
+			ICPPTemplateParameterMap parameterMap = ati.getTemplateParameterMap();
+			Stream.of(parameterMap.getAllParameterPositions()).forEach(it -> {
+				ICPPTemplateArgument arg = parameterMap.getArgument(it);
+				if (arg != null) {
+					if (arg.isTypeValue())
+						templateParameters.append(resolveType(arg.getTypeValue(), origin));
+					else
+						templateParameters.append(resolveType(arg.getNonTypeEvaluation().getType(), origin));
+				} else {
+					// TODO: Rodin, why is this empty?
```

Comment:
This looks like the same code as `resolveICPPClassSpecialization`

Reply:
it is! I cloned it.